runtime/runtime_test.go

```diff
@@ -4460,139 +4462,251 @@ func TestRuntimeBlock(t *testing.T) {
+
+		runValidCaseWithModulo := func(t *testing.T, ty sema.Type) {
+			// build a big Int from the modulo buffer, with the required `ty` size
+			// big.Int are used as they cover all the tested types including the small ones (UInt8 ..)
+			modulo := new(big.Int).SetBytes(moduloBuffer[:typeToBytes[ty]])
+
+			value, err := executeScript(ty, modulo.String(), readCryptoRandom)
+			require.NoError(t, err)
+			// convert `value` to big Int for comparison
+			valueBig, ok := new(big.Int).SetString(value.String(), 10)
+			require.True(t, ok)
+			// check that modulo > value
+			require.True(t, modulo.Cmp(valueBig) == 1)
+		}
+		testAllTypes(t, runValidCaseWithModulo)
+	})
+
+	// test valid edge cases of the value modulo - test all types
+	t.Run("script with modulo edge cases all types", func(t *testing.T) {
+		// case where modulo is the max value of the type
+		runValidCaseWithMaxModulo := func(t *testing.T, ty sema.Type) {
+
+			// set modulo to the max value of the type: (1 << bitSize) - 1
+			// big.Int are used as they cover all the tested types including the small ones (UInt8 ..)
+			bitSize := typeToBytes[ty] << 3
+			one := big.NewInt(1)
+			modulo := new(big.Int).Lsh(one, uint(bitSize))
+			modulo.Sub(modulo, one)
+
+			value, err := executeScript(ty, modulo.String(), readCryptoRandom)
+			require.NoError(t, err)
+			// convert `value` to big Int for comparison
+			valueBig, ok := new(big.Int).SetString(value.String(), 10)
+			require.True(t, ok)
+			// check that modulo > value
+			require.True(t, modulo.Cmp(valueBig) == 1)
+		}
 
-			require.Equal(t, getFixedSizeUnsignedIntegerForSemaType(ty), value)
+		// case where modulo is 1 and expected value in 0
+		runValidCaseWithOneModulo := func(t *testing.T, ty sema.Type) {
+			// set modulo to 1
+			value, err := executeScript(ty, "1", readCryptoRandom)
+			require.NoError(t, err)
+			// check that value is zero
+			require.True(t, value.String() == "0")
+		}
+
+		t.Run("max modulo", func(t *testing.T) {
+			testAllTypes(t, runValidCaseWithMaxModulo)
 		})
-	}
+		t.Run("one modulo", func(t *testing.T) {
+			testAllTypes(t, runValidCaseWithOneModulo)
+		})
+	})
+
+	// function should error if zero is used as modulo - test all types
+	t.Run("script with zero modulo", func(t *testing.T) {
+		runCaseWithZeroModulo := func(t *testing.T, ty sema.Type) {
+			// set modulo to "0"
+			_, err := executeScript(ty, "0", readCryptoRandom)
+			assertUserError(t, err)
+			require.ErrorContains(t, err, stdlib.ZeroModuloError.Error())
+		}
+		testAllTypes(t, runCaseWithZeroModulo)
+	})
+
+	// sanity statistical test to make sure the random numbers less than modulo
+	// are uniform in [0,modulo-1].
+	// The test requires the original random source (here `crypto/rand`)
+	// to be uniform.
+	// The test uses the same small values for all types: one is a power of 2
+	// and the other is not.
+	t.Run("basic uniformity with modulo", func(t *testing.T) {
+		// skipped in CI because the test is slow
+		t.Skip()
```

Comment:
```suggestion
		if testing.Short() {
			t.Skip()
		}
```

Reply:
nice!

Reply:
@turbolent once I implement the suggestion, I noticed that CI hits a 10mn timeout. Can that timeout be increased?

Reply:
I'll revert https://github.com/onflow/cadence/pull/2986/commits/e0f07e3901eb8ce8bf227bee414004bccd95216c and https://github.com/onflow/cadence/pull/2986/commits/42271b5b35d71dcfe5bbe080952aa908485c9bdc once we overcome the CI constraint

Reply:
Can the test be reduced to take less than 10 min? e.g. 1-2 min?

Reply:
It actually takes less than a minute on my laptop, I'm surprised CI is more than 10x slower. I'll see what I can do, otherwise we may need to disable the test on CI.

Reply:
Btw I saw you added extra parallelizations in https://github.com/onflow/cadence/pull/2986/commits/2aaa16e5fd3ce8546100c22d99a3573c60083003. To clarify:
- I think sub tests are already parallelized per type here https://github.com/onflow/cadence/blob/2aaa16e5fd3ce8546100c22d99a3573c60083003/runtime/runtime_test.go#L4539
- only [one test](https://github.com/onflow/cadence/blob/2aaa16e5fd3ce8546100c22d99a3573c60083003/runtime/runtime_test.go#L4708) is causing slowness, the others are quick. This is why I disabled that test in CI in my initial PR submission, before you suggested to check `testing.Short()`.

Reply:
I'm also surprised each sub-test for each type takes much longer on CI than on my machine (seconds vs minutes).
If tests are disabled on CI, they will unlikely be run on developer machines, and when code that is tested by it is changed, CI will not run it.

Instead it is better to run as many tests on CI as possible, even if they take a a bit longer. If a developer wants to skip those long tests on their machine, they can do so using `go test -short`.

We can up the default timeout of 10min if needed, but it seems odd why the tests take so long in the first place.

Regarding the parallelization: I also assumed that it was "sufficient" to just parallelize the innermost tests (the one for each type, which, as you pointed out above, was already parallelized).
However, when the outer test scheduling the tests for each type was not itself parallelized, the test framework waited until the first `testTypes` was complete, before scheduling the second `testTypes`.
We want as much parallelism as possible, so also need to indicate that the two `testTypes` can be run at the same time.

It makes a difference on machines with many cores, but not on CI. We might want to also consider upping the CI runners like we do on flow-go

Reply:
I agree with running the heavy checks in CI, and keeping the light version for `go test -short`. However there is a limitation for "heavy" if all CI tests have to wrap up under the timeout. Btw I noticed that succeeding CI tests in other PRs currently run in 9-10mn, which is already very close to the timeout.

I've refactored the statistical test in  https://github.com/onflow/cadence/pull/2986/commits/395fb0a22d9b1c0694aa09a94ac6fe7554cdcfc9 so that the 85000 randoms needed from scripts are returned in batches of 5000, that way we minimize the number of times a script is run, unfortunately we still hit the timeout (and I'm not even enabling all types). 
I see two options to move forward: 
- increase the 10mn timeout
- Or keep the statistical test in the code base but disable it in CI till we up the CI runners.

Reply:
I'll try to improve the test further or disable it on CI